packages/wrangler/src/type-generation/runtime/log-runtime-types-message.ts

```diff
@@ -0,0 +1,79 @@
+import { existsSync, readFileSync } from "fs";
+import { logger } from "../../logger";
+
+/**
+ * Constructs a comprehensive log message for the user after generating runtime types.
+ */
+export function logRuntimeTypesMessage(outFile: string, tsconfigPath: string) {
+	const existingTypes = readExistingTypes(tsconfigPath);
+	const isWorkersTypesInstalled = workersTypesEntryExists(existingTypes);
+	const updatedTypesString = buildUpdatedTypesString(existingTypes, outFile);
+
+	const message = isWorkersTypesInstalled
+		? `\n📣 Replace the existing "@cloudflare/workers-types" entry with the generated types path:`
+		: "📣 Add the generated types to the types array in your tsconfig.json:";
+
+	logger.info(`
+✨ Runtime types written to ${outFile}
+${message}
+
+{
+	"compilerOptions": {
+		"types": ${updatedTypesString}
+	}
+}
+    `);
+	if (isWorkersTypesInstalled) {
+		logger.info('📣 You can now uninstall "@cloudflare/workers-types".');
+	}
+	logger.info(
+		'📣 To get Node.js typings, install with "npm i --save-dev @types/node".'
+	);
+	logger.info(
+		"📣 Remember to run 'wrangler types --x-with-runtime' again if you change 'compatibility_date' or 'compatibility_flags' in your wrangler.toml.\n"
+	);
+}
+
+/**
+ * Constructs a string representation of the existing types array with the new types path appended to.
+ * It removes any existing types that are no longer relevant.
+ */
+function buildUpdatedTypesString(
+	types: string[],
+	newTypesPath: string
+): string {
+	const updatedTypesArray = types
+		.filter(
+			(type) =>
+				!type.startsWith("@cloudflare/workers-types") &&
+				!type.includes(".wrangler/types/runtime")
+		)
+		.concat([newTypesPath]);
+
+	return JSON.stringify(updatedTypesArray);
+}
+
+/**
+ * Attempts to read the tsconfig.json at the current path.
+ */
+function readExistingTypes(tsconfigPath: string): string[] {
+	if (!existsSync(tsconfigPath)) {
+		return [];
+	}
+
+	try {
+		const tsconfig = JSON.parse(readFileSync(tsconfigPath, "utf-8"));
```

Comment:
tsconfig files are JSONC, so this may not work. You probably want the [`parseJSONC` utility function](https://github.com/cloudflare/workers-sdk/tree/main/packages/wrangler/src/parse.ts#L179)

Reply:
Thanks, good call.